common/state-machine-lib/src/main/java/org/eclipse/dataspaceconnector/common/statemachine/StateMachine.java

```diff
@@ -0,0 +1,149 @@
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.common.statemachine;
+
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.retry.WaitStrategy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.lang.String.format;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Handles a loop that processes entities continuously.
+ * On every iteration it runs all the set processors sequentially,
+ * applying a wait strategy in the case no entities are processed on the iteration.
+ */
+public class StateMachine {
+
+    private final List<EntitiesProcessor> processors = new ArrayList<>();
+    private final ExecutorService executor;
+    private final AtomicBoolean active = new AtomicBoolean();
+    private final WaitStrategy waitStrategy;
+    private final Monitor monitor;
+    private final String name;
+    private int shutdownTimeout = 10;
+
+    private StateMachine(String name, Monitor monitor, WaitStrategy waitStrategy) {
+        this.name = name;
+        this.monitor = monitor;
+        this.waitStrategy = waitStrategy;
+        this.executor = Executors.newSingleThreadExecutor(r -> {
+            var thread = Executors.defaultThreadFactory().newThread(r);
+            thread.setName("StateMachine-" + name);
+            return thread;
+        });
+    }
+
+    /**
+     * Start the loop that will run processors until it's stopped
+     *
+     * @return a future that will complete when the loop stops
+     */
+    public Future<?> start() {
+        active.set(true);
+        return executor.submit(loop());
+    }
+
+    /**
+     * Stop the loop gracefully
+     *
+     * @return a future that will complete when the loop is fully stopped.
+     *         The content of the future will be true if stop happened before the timeout, false elsewhere.
+     */
+    public CompletableFuture<Boolean> stop() {
+        active.set(false);
+
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                return executor.awaitTermination(shutdownTimeout, SECONDS);
+            } catch (InterruptedException e) {
+                monitor.severe(format("StateMachine [%s] await termination failed", name), e);
+                return false;
+            }
+        });
+    }
+
+    /**
+     * Tells if the loop is active and running
+     *
+     * @return true if it's active, false otherwise
+     */
+    public boolean isActive() {
+        return active.get();
+    }
+
+    private Runnable loop() {
+        return () -> {
+            while (active.get()) {
```

Comment:
nit: Just FYI, for #790, we will have to change this to use scheduled executor rather than endless loop.

Reply:
Yes, I saw it, so it will be easier to that since all this logic will be in a single class. :+1: